arch/arm/src/s32k3xx/s32k3xx_serial.c

```diff
@@ -1091,99 +1435,295 @@ static const struct uart_ops_s g_uart_ops =
   .rxint          = s32k3xx_rxint,
   .rxavailable    = s32k3xx_rxavailable,
 #ifdef CONFIG_SERIAL_IFLOWCONTROL
-  .rxflowcontrol  = NULL,
+  .rxflowcontrol  = s32k3xx_rxflowcontrol,
 #endif
   .send           = s32k3xx_send,
   .txint          = s32k3xx_txint,
   .txready        = s32k3xx_txready,
   .txempty        = s32k3xx_txempty,
 };
+#endif
+
+#if defined(SERIAL_HAVE_RXDMA) && defined(SERIAL_HAVE_TXDMA)
+static const struct uart_ops_s g_lpuart_rxtxdma_ops =
+{
+  .setup          = s32k3xx_dma_setup,
+  .shutdown       = s32k3xx_dma_shutdown,
+  .attach         = s32k3xx_attach,
+  .detach         = s32k3xx_detach,
+  .ioctl          = s32k3xx_ioctl,
+  .receive        = s32k3xx_dma_receive,
+  .rxint          = s32k3xx_dma_rxint,
+  .rxavailable    = s32k3xx_dma_rxavailable,
+#ifdef CONFIG_SERIAL_IFLOWCONTROL
+  .rxflowcontrol  = s32k3xx_rxflowcontrol,
+#endif
+  .send           = s32k3xx_send,
+  .txint          = s32k3xx_dma_txint,
+  .txready        = s32k3xx_txready,
+  .txempty        = s32k3xx_txempty,
+  .dmatxavail     = s32k3xx_dma_txavailable,
+  .dmasend        = s32k3xx_dma_send,
+};
+#endif
+
+#if !defined(SERIAL_HAVE_ONLY_DMA) && defined(SERIAL_HAVE_RXDMA)
+static const struct uart_ops_s g_lpuart_rxdma_ops =
+{
+  .setup          = s32k3xx_dma_setup,
+  .shutdown       = s32k3xx_dma_shutdown,
+  .attach         = s32k3xx_attach,
+  .detach         = s32k3xx_detach,
+  .ioctl          = s32k3xx_ioctl,
+  .receive        = s32k3xx_dma_receive,
+  .rxint          = s32k3xx_dma_rxint,
+  .rxavailable    = s32k3xx_dma_rxavailable,
+#ifdef CONFIG_SERIAL_IFLOWCONTROL
+  .rxflowcontrol  = s32k3xx_rxflowcontrol,
+#endif
+  .send           = s32k3xx_send,
+  .txint          = s32k3xx_txint,
+  .txready        = s32k3xx_txready,
+  .txempty        = s32k3xx_txempty,
+};
+#endif
+
+#if !defined(SERIAL_HAVE_ONLY_DMA) && defined(SERIAL_HAVE_TXDMA)
+static const struct uart_ops_s g_lpuart_txdma_ops =
+{
+    .setup          = s32k3xx_dma_setup,
+    .shutdown       = s32k3xx_dma_shutdown,
+    .attach         = s32k3xx_attach,
+    .detach         = s32k3xx_detach,
+    .ioctl          = s32k3xx_ioctl,
+    .receive        = s32k3xx_receive,
+    .rxint          = s32k3xx_rxint,
+    .rxavailable    = s32k3xx_rxavailable,
+  #ifdef CONFIG_SERIAL_IFLOWCONTROL
+    .rxflowcontrol  = s32k3xx_rxflowcontrol,
+  #endif
+    .send           = s32k3xx_send,
+    .txint          = s32k3xx_dma_txint,
+    .txready        = s32k3xx_txready,
+    .txempty        = s32k3xx_txempty,
+    .dmatxavail     = s32k3xx_dma_txavailable,
+    .dmasend        = s32k3xx_dma_send,
+};
+#endif
```

Comment:
Question: since these three variables are basically the same, could it be possible to use #ifdefs inside the struct initialization to setup ".rxint", ".txint", ".dmatxavail" and ".dmasend" directly?
I agree that for readability doing the way you did is better, I'm thinking more about code reduction.

Reply:
Possibly, but I think I will not be readable.

Reply:
hmmm...Maybe not... You can have one port with only TX and another with only RX DMA and then one with none and one with both. So the tables can not be reduced.